### DISCUSSION TITLE
Why Does Neo Tree Show Different Directories for CWD and Root Dir in a Monorepo?

### DISCUSSION THREAD

**Discussion by Demianeen** [LazyVim/LazyVim#2150] [Q&A] [answered]
Hello everyone,

I started to use LazyVim with Neo Tree in a monorepo setup, and I'm encountering an issue with how the file explorer displays directories. Our monorepo has only one Git repository at the top. When I open Neo Tree using the 'root dir' option, it only shows one package in the monorepo, even though I opened Vim at the repo's root. In contrast, when I use the 'cwd' (current working directory) option, Neo Tree opens in the root of the entire repo as expected.

I asked neo-tree community and the neo-tree maintainer [answered](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/1254#discussioncomment-7779721):
> I think those commands/options are something specific to LazyVim.
> Neo-tree will just use the current working directory by default, unless the command is run with the dir= option, in which case it will use that dir as the root. You can find information on how to use the Neotree command on the README page: [nvim-neo-tree/neo-tree.nvim#the-neotree-command](https://github.com/nvim-neo-tree/neo-tree.nvim?tab=readme-ov-file#the-neotree-command)

Is there a specific configuration on the Lazyvim site or setting that determines how Neo Tree identifies the root directory in a monorepo setup, especially when there's only one Git repository at the top level?

Any insights or suggestions to resolve this issue would be greatly appreciated!

**Comment by dpetka2001** [comment, marked as answer]
LazyVim uses `lazyvim/util/root` to determine the `root_dir` of a buffer. First it takes into account the LSP server that is attached to the buffer and returns that as `root_dir` and if it doesn't find anything it falls back to `vim.loop.cwd`, which returns the current working directory. 

Usually the workflow would be to open Neovim at the root of your project and then use `<leader><space>` to show the contents of the LSP `root_dir` attached to the buffer. If your monorepo includes subfolder, which have a file that is used as a pattern by the LSP server to determine the `root_dir`, then that directory is returned to begin the search when using `<leader><space>`. If you want to do searches from the root of your project downwards, then you use `<leader>fF`. The same applies for the Neo-tree keymaps (`<leader>e` and `<leader>E`).

If you don't like this behavior, you can change `vim.g.root_spec` to include what you want. Changing this to `vim.g.root_spec = { "cwd" }` in your `lua/config/options.lua` file will get rid of LazyVim's default behavior and always return the current working directory for your searches. It's up to you which choice to make.

**Reply by A-caibird** [reply]
This is amazing! It solves a problem that has been bothering me for a long time.

**Comment by cerebrl** [comment]
For me, the LSP configuration for some of LazyVim's extras were causing some unwanted root directory detection that I didn't like. After a lot of experimentation, I found that altering how the LSPs detects the root dir was the most helpful. For languages like JavaScript/TypeScript, there are a lot of LSPs all trying to detect the root in different ways, so I had to change 3 of the servers.

Here's my config, if you'd like to see how I did it. I mostly just had the servers use the `.git` file as the signal for the root dir: https://github.com/cerebrl/nvim/blob/main/lua/plugins/lsp.lua#L35. Hope that helps.

Correction: I may have been wrong, as it's not LazyVim's configuration that was causing my issue, but it was actually `nvim-lspconfig`. Here's the TS config as an example: https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#tsserver. I guess I was overwriting that, rather than the config in LazyVim.

**Reply by dpetka2001** [reply]
Yes, I would also suggest to do that. `:LazyRoot` will show you the available `roots` that LazyVim detects and you can change the LSP server accordingly. It's just that a lot of people complained about this behavior in the past, that I almost took it for granted that OP would also want to do the same. But it's definitely best to lay out all possible solutions and let OP decide what's best for him.
